text subs: use generic font family names

jstebbins opened the pull request:
Allows the platform more flexibility in choosing the "best" font for the
platform.

jstebbins committed:
text subs: use generic font family names

Allows the platform more flexibility in choosing the "best" font for the
platform.

jstebbins commented:
See https://github.com/HandBrake/HandBrake/issues/471 for a discussion of this

bradleysepos commented:
Looks good.

jstebbins commented:
Needs to be tested with CoreText and DirectWrite once we get the fontconfig stuff ironed out

jstebbins commented:
FYI, on Linux this results in DejaVuSansMono for monospace which looks decent.  

For some reason, my mingw build with fontconfig fails to burn in subtitles.  It never shows the line in the log where it selects a font.  e.g.
> [10:01:53] [ass] fontselect: (monospace, 400, 0) -> /usr/share/fonts/dejavu/DejaVuSansMono.ttf, 0, DejaVuSansMono

When I build astiob directwrite PR it selects CourierNew which is what we were trying to avoid because the font is thin and has serifs.  So I'm not sure monospace is the ultimate solution yet. On windows 10 there is also Lucida Console which looks very similar to DejaVuSansMono on linux.  Does anyone know if this is generally available on all windows versions we support?

So I think we are going to want to special case the windows font name at least.  Someone also needs to test what OSX selects.  If OSX is selecting an ugly or unreadable font, we might want to check Monaco.

bradleysepos commented:
Good info.

Mac should be fine but I'll have to run a test to be sure. Sample file with CC available?

Lucida Console should be fine on Windows; Wikipedia says its inclusion dates to 98.

jstebbins commented:
Example CC file
https://john.stebbins.name/owncloud/index.php/s/unxOz3p9TpbjJTu

jstebbins committed:
rendersub: make font configurable per platform

Use Lucida Console for mono font on windows since it chooses ugly,
difficult to read Courier New when using monospace family name.

bradleysepos commented:
[14:29:54] [ass] fontselect: (Courier New, 400, 0) -> /Library/Fonts/Courier New.ttf, -1, CourierNewPSMT

Though I am a Monaco fan (I write and program in Monaco 11 pt *anti-aliased*, oddly enough), I believe it preferable to use Courier for CC on Mac.

bradleysepos commented:
And I tested the wrong branch...

bradleysepos commented:
Okay, it looks like libass is doing The Right Thing™ with CoreText:

    [14:39:49] [ass] fontselect: (monospace, 400, 0) -> /System/Library/Fonts/Courier.dfont, -1, Courier

So it would seem we only need to alias `HB_FONT_MONO` to Lucida Console on Windows.

jstebbins commented:
Cool.  Since I've tested this change on top of the directwrite PR, I think this is safe to commit.  Any objections?

WhosAsking commented:
Is there anything preventing you from attaching a free-to-use font (like Deja Vu Sans Mono) to an MKV file so that a closed-captioning font is always available no matter what the system?

bradleysepos commented:
Read my response here: https://github.com/HandBrake/HandBrake/issues/471#issuecomment-321174411

If you're going to resurrect 5 month old threads, consider avoiding duplicate posts.
